ui/app/components/variable-form.js

```diff
@@ -326,4 +329,53 @@ export default class VariableFormComponent extends Component {
       this.args.model.pathLinkedEntities?.task
     );
   }
+
+  //#region Unsaved Changes Confirmation
+
+  @computed(
+    'args.model.{keyValues,path}',
+    'keyValues.@each.{key,value}',
+    'path'
+  )
+  get hasUserModifiedAttributes() {
+    const compactedBasicKVs = this.keyValues
+      .map((kv) => ({ key: kv.key, value: kv.value }))
+      .filter((kv) => kv.key || kv.value);
+    const compactedPassedKVs = this.args.model.keyValues.filter(
+      (kv) => kv.key || kv.value
+    );
+    const unequal =
+      !isEqual(compactedBasicKVs, compactedPassedKVs) ||
+      !isEqual(this.path, this.args.model.path);
+    return unequal;
+  }
+
+  addExitHandler() {
+    this.router.on('routeWillChange', this, this.confirmExit);
+  }
+
+  removeExitHandler() {
+    this.router.off('routeWillChange', this, this.confirmExit);
+  }
+
+  confirmExit(transition) {
+    if (transition.isAborted) return;
+
+    if (this.hasUserModifiedAttributes) {
+      if (
+        !confirm(
+          'Your variable has unsaved changes. Are you sure you want to leave?'
+        )
```

Comment:
`question`: Wouldn't this nested conditional always resolve to a `falsy` value?

Taking a look at the event handler `confirm` it doesn't have a `return` and would thus return `void` a `falsy` value every time.

What are you trying to guard against? Maybe that the confirmation action has successfully dispatched... in which case the only thing that can confirm that would be being in the `idle` state.

Reply:
Confirm returns whatever the user selects. If "OK", it comes back true. If "Cancel", false.